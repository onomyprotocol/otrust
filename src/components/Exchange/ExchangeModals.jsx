import React, { useState } from 'react';
import Modal from 'react-modal';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft, faExchangeAlt } from '@fortawesome/free-solid-svg-icons';

import { responsive } from 'theme/constants';
import { SellBtn, ExchangeButton, Sending, ExchangeInput, MaxBtn, Receiving, ReceivingValue } from './exchangeStyles';

const ModalTrigger = styled.div`
  display: none;

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 24px;
  }
`;

const ExchangeModalWrapper = styled.div`
  width: 100%;
  height: 100%;

  background-color: ${props => props.theme.colors.bgDarkest};

  header {
    padding: 20px 20px 0;

    background-color: ${props => props.theme.colors.bgNormal};
  }
`;

const ModalHeader = styled.header`
  display: flex;
  align-items: center;

  h6 {
    margin: 0 auto;

    font-size: 16px;
  }
`;

const FormWrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 12px;

  padding: 56px 20px;
`;

const HeaderInfoItem = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;

  @media screen and (max-width: ${responsive.laptop}) {
    gap: 5px;
  }

  & + & {
    margin-left: 56px;

    @media screen and (max-width: ${responsive.laptop}) {
      margin-left: 32px;
    }
  }

  > strong {
    color: ${props => props.theme.colors.textThirdly};
    font-weight: 400;

    @media screen and (max-width: ${responsive.smartphoneLarge}) {
      font-size: 10px;
    }
  }
`;

const HeaderInfoItemValue = styled.div`
  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    display: flex;
    flex-direction: column;

    gap: 5px;
  }

  > strong {
    margin-right: 12px;

    color: ${props => props.theme.colors.textPrimary};
    font-family: 'Bebas Neue', sans-serif;
    font-size: 24px;

    @media screen and (max-width: ${responsive.laptop}) {
      font-size: 20px;
    }

    @media screen and (max-width: ${responsive.smartphoneLarge}) {
      font-size: 18px;
    }
  }
`;

const ModalInfo = styled.div`
  display: flex;
  justify-content: space-between;

  padding: 32px 20px;

  background-color: ${props => props.theme.colors.bgNormal};
`;

const ModalBtn = styled.button`
  width: 44px;
  height: 44px;

  border-radius: 8px;
  border: none;
  background-color: ${props => props.theme.colors.bgHighlightBorder};

  color: #84809a;

  cursor: pointer;
`;

const modalOverride = {
  content: {
    padding: 0,
    border: 'none',
    borderRadius: 0,

    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
  },
};

export default function ExchangeModals() {
  const [buyModalOpen, setBuyModalOpen] = useState(false);
  const [sellModalOpen, setSellModalOpen] = useState(false);

  return (
    <ModalTrigger>
      <ExchangeButton onClick={() => setBuyModalOpen(true)}>Buy NOM</ExchangeButton>

      <SellBtn onClick={() => setSellModalOpen(true)}>Sell NOM</SellBtn>

      <Modal isOpen={buyModalOpen} style={modalOverride} onRequestClose={() => setBuyModalOpen(false)}>
        <ExchangeModalWrapper>
          <ModalHeader>
            <ModalBtn onClick={() => setBuyModalOpen(false)}>
              <FontAwesomeIcon icon={faChevronLeft} />
            </ModalBtn>
            <h6>Buy NOM</h6>
            <ModalBtn
              onClick={() => {
                setSellModalOpen(!sellModalOpen);
                setBuyModalOpen(!buyModalOpen);
              }}
            >
              <FontAwesomeIcon icon={faExchangeAlt} />
            </ModalBtn>
          </ModalHeader>

          <ModalInfo>
            <HeaderInfoItem>
              <strong>Eth Balance</strong>
              <HeaderInfoItemValue>
                <strong>12.12321</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM Balance</strong>
              <HeaderInfoItemValue>
                <strong>10,432.22</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM / USDT</strong>
              <HeaderInfoItemValue>
                <strong>$10.12</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM / ETH</strong>
              <HeaderInfoItemValue>
                <strong>0.07</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
          </ModalInfo>

          <FormWrapper>
            <strong>Buy NOM</strong>
            <Sending>
              <strong>I'm sending</strong>
              <ExchangeInput type="text" value="0.15 ETH" />
              <MaxBtn>Max</MaxBtn>
            </Sending>
            <Receiving>
              <strong>I'm receiving</strong>
              <ReceivingValue>123 NOM</ReceivingValue>
            </Receiving>
            <div>
              <ExchangeButton>Buy NOM</ExchangeButton>
            </div>
          </FormWrapper>
        </ExchangeModalWrapper>
      </Modal>

      <Modal isOpen={sellModalOpen} style={modalOverride} onRequestClose={() => setSellModalOpen(false)}>
        <ExchangeModalWrapper>
          <ModalHeader>
            <ModalBtn onClick={() => setSellModalOpen(false)}>
              <FontAwesomeIcon icon={faChevronLeft} />
            </ModalBtn>
            <h6>Sell NOM</h6>
            <ModalBtn
              onClick={() => {
                setSellModalOpen(!sellModalOpen);
                setBuyModalOpen(!buyModalOpen);
              }}
            >
              <FontAwesomeIcon icon={faExchangeAlt} />
            </ModalBtn>
          </ModalHeader>

          <ModalInfo>
            <HeaderInfoItem>
              <strong>Eth Balance</strong>
              <HeaderInfoItemValue>
                <strong>12.12321</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM Balance</strong>
              <HeaderInfoItemValue>
                <strong>10,432.22</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM / USDT</strong>
              <HeaderInfoItemValue>
                <strong>$10.12</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
            <HeaderInfoItem>
              <strong>NOM / ETH</strong>
              <HeaderInfoItemValue>
                <strong>0.07</strong>
              </HeaderInfoItemValue>
            </HeaderInfoItem>
          </ModalInfo>

          <FormWrapper>
            <strong>Sell NOM</strong>
            <Sending>
              <strong>I'm sending</strong>
              <ExchangeInput type="text" value="2529 NOM" />
              <MaxBtn>Max</MaxBtn>
            </Sending>
            <Receiving>
              <strong>I'm receiving</strong>
              <ReceivingValue>123 ETH</ReceivingValue>
            </Receiving>
            <div>
              <SellBtn>Sell NOM</SellBtn>
            </div>
          </FormWrapper>
        </ExchangeModalWrapper>
      </Modal>
    </ModalTrigger>
  );
}
