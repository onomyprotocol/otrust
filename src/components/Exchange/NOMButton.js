import React, { useEffect } from 'react';

import { useExchange } from 'context/exchange/ExchangeContext';
import { useChain } from 'context/chain/ChainContext';
import { SellBtn } from './exchangeStyles';

export default function NOMButton({ onBid, onApprove }) {
  const { weakBalance, NOMallowance } = useChain();

  const { bidAmount, bidDenom, input, weak } = useExchange();

  useEffect(() => {
    console.log('Bid Amount: ', bidAmount);
    console.log('NOM Allowance: ', NOMallowance);
  }, [NOMallowance, bidAmount]);

  return (
    <>
      {bidDenom === 'strong' ? (
        <SellBtn> Input Value </SellBtn>
      ) : bidAmount.lte(weakBalance) ? (
        input === '' ? (
          <SellBtn> Input Value </SellBtn>
        ) : NOMallowance.gt(bidAmount) ? (
          <SellBtn onClick={onBid}>Sell {weak}</SellBtn>
        ) : (
          <SellBtn onClick={onApprove}>Approve</SellBtn>
        )
      ) : (
        <SellBtn> Low {weak} Balance </SellBtn>
      )}
    </>
  );
}
