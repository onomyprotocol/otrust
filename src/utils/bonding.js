const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
