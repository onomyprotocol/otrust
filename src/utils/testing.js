import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';

import { dark } from 'Theme/theme';

export const renderWithTheme = (Component, rest) => {
  return render(
    <ThemeProvider theme={dark}>
      <Component {...rest} />
    </ThemeProvider>,
  );
};
