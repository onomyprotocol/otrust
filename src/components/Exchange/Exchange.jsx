import React, { useCallback, useState, useEffect } from "react";
import { useAsyncFn } from "lib/use-async-fn";
import { BigNumber } from 'bignumber.js'
import { format18, parse18 } from 'utils/math'

import {
  ExchangeWrapper,
  ExchangeItem,
  Sending,
  Receiving,
  ExchangeInput,
  MaxBtn,
  ReceivingValue,
  SellBtn,
  ExchangeButton,
} from "./exchangeStyles";

import ConfirmTransactionModal from '../Modals/ConfirmTransactionModal';
import { Dimmer } from "components/UI/Dimmer";

import ExchangeModals from "./ExchangeModals";
import { useModal, useUpdateModal } from 'context/ModalContext'
import { useSwap, useUpdateSwap } from "context/SwapContext";
import { useChain, useUpdateChain } from "context/chain/ChainContext";

import TransactionCompletedModal from "components/Modals/TransactionCompletedModal";
import OnomyConfirmationModal from "components/Modals/OnomyConfirmationModal";
import TransactionFailedModal from "components/Modals/TransactionFailedModal";
import PendingModal from "components/Modals/PendingModal";

export default function Exchange() {
  const {
    approveModal,
    completedModal,
    confirmModal,
    failedModal,
    pendingModal
  } = useModal();

  const {
    setApproveModal,
    setCompletedModal,
    setConfirmModal,
    setFailedModal,
    setPendingModal
  } = useUpdateModal()

  const { 
    swapBuyAmount, 
    swapBuyResult,
    swapBuyValue,
    swapSellAmount, 
    swapSellResult,
    swapSellValue, 
    swapDenom 
  } = useSwap();
  
  const { 
    setSwapBuyAmount, 
    setSwapBuyResult, 
    setSwapBuyValue, 
    setSwapSellAmount, 
    setSwapSellResult,
    setSwapSellValue, 
    setSwapDenom 
  } = useUpdateSwap();

  const { 
    bondContract, 
    NOMallowance, 
    NOMcontract, 
    ETHbalance, 
    NOMbalance, 
    pendingTx 
  } = useChain();
  
  const { setPendingTx } = useUpdateChain();
  
  const [completedAmount, setCompletedAmount] = useState(null);
  const [completedResult, setCompletedResult] = useState(null);
  const [slippage, setSlippage] = useState(1);
  const [previousTx, setPreviousTx] = useState(null);
 
  const onBuyNOMTextChange = useCallback(
    (evt) => {
      evt.preventDefault()
      setSwapBuyValue(evt.target.value)
      setSwapSellAmount('')
      setSwapSellResult('')
      setSwapSellValue('')
      setSwapDenom('ETH')
      if (BigNumber.isBigNumber(evt.target.value)) {
        setSwapBuyAmount('')
        setSwapBuyResult('')
      } else {
        try {
          setSwapBuyAmount(
            parse18(
              new BigNumber(
                evt.target.value
                )
                )
                )
                  
        } catch {
          setSwapBuyAmount(new BigNumber(0))
        }
        
      }
    },
    [
      setSwapBuyAmount,
      setSwapBuyResult,
      setSwapBuyValue,
      setSwapDenom, 
      setSwapSellAmount,
      setSwapSellResult
    ]
  );
  
  const onSellNOMTextChange = useCallback(
    (evt) => {
      evt.preventDefault()
      setSwapSellValue(evt.target.value)
      setSwapBuyAmount('')
      setSwapBuyResult('')
      setSwapDenom('NOM')

      if (!evt.target.value) {
        setSwapSellAmount('')
        setSwapSellResult('')
      } else {
        try {
          setSwapSellAmount(parse18(new BigNumber(evt.target.value)))
        } catch {
          setSwapSellAmount(new BigNumber(0))
        }
        
      }
    },
    [
      setSwapBuyAmount, 
      setSwapBuyResult,
      setSwapDenom,
      setSwapSellAmount,
      setSwapSellResult,
      setSwapSellValue
    ]
  );

  const submitTrans = useCallback(
    async (denom) => {
      if (!swapBuyAmount && !swapSellAmount) return;
      try {
        let tx;
        if (denom === "ETH") {
          tx = await bondContract.buyNOM(
            swapBuyResult,
            slippage * 100,
            { value: swapBuyAmount }
          );
        } else {
          tx = await bondContract.sellNOM(
            swapSellAmount,
            swapSellResult,
            slippage * 100,
          );
        }
        setPendingTx(tx);
        setConfirmModal('');
        setCompletedAmount(denom === 'ETH' ? swapBuyAmount : swapSellAmount);
        setCompletedResult(denom === 'ETH' ? swapBuyResult : swapSellResult);
        setPendingModal(true);
        setSwapBuyAmount("");
        setSwapSellAmount("");
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e.code, e.message.message);
        // alert(e.message)
        setFailedModal(e.code + '\n' + e.message.slice(0,80) + '...')
        // setSwapBuyAmount(swapBuyAmount);
      }
    },
    [
      bondContract,
      setCompletedAmount,
      setCompletedResult,
      setConfirmModal,
      setFailedModal,
      setPendingModal,
      setPendingTx,
      setSwapSellAmount,
      setSwapBuyAmount,
      slippage,
      swapBuyAmount,
      swapBuyResult,
      swapSellAmount,
      swapSellResult
    ]
  );

  useEffect(() => {
    if (pendingTx) {
      pendingTx.wait().then(() => {
        setPreviousTx(pendingTx);
        setCompletedModal(swapDenom);
        setPendingTx(null);
        setPendingModal(false);
      })
    }
  }, [
    pendingTx,
    setCompletedModal,
    setPendingModal,
    setPendingTx,
    swapBuyAmount,
    swapBuyResult,
    swapDenom,
    swapSellAmount,
    swapSellResult,
  ])

  const onBuy = () => {
    setSwapDenom('ETH');
    setConfirmModal('ETH');
  }

  const onSell = () => {
    setSwapDenom('NOM');
    setConfirmModal('NOM');
  }
  
  const onApprove = async (value) => {
    if(value <= NOMbalance) {
      try {
        setApproveModal(false);
        setPendingModal(true);
        setSwapDenom('APPROVE')
        let tx = await NOMcontract.increaseAllowance(
          bondContract.address,
          value
        );
        setPendingTx(tx);
      } catch (e) {
        // eslint-disable-next-line no-console
        // console.error(e.code, e.message.message);
        // alert(e.message)
        setFailedModal(e.code + '\n' + e.message.slice(0,80) + '...')
        // setSwapBuyAmount(swapBuyAmount);
      }    
    } else {
      setFailedModal('NOM Balance too low')
    }
  } 
  
  

  const [onSubmit, error] = useAsyncFn(submitTrans);

  const onEthMax = () => {
    setSwapBuyAmount(ETHbalance)
  }

  const onNOMMax = () => {
    setSwapSellAmount(NOMbalance)
  }
  return (
    <ExchangeWrapper>
      <ExchangeModals />
      {confirmModal && 
        <Dimmer>
          <ConfirmTransactionModal
            closeModal={() => setConfirmModal('')}
            type={confirmModal}
            amount={confirmModal === 'ETH' ? swapBuyAmount : swapSellAmount}
            result={confirmModal === 'ETH' ? swapBuyResult : swapSellResult}
            onConfirm={() => 
              onSubmit(confirmModal)
            }
            setSlippage={setSlippage}
            slippage={slippage}
          />
        </Dimmer>
      }
        {/* <TransactionCompletedModal /> */}
      {
        approveModal && 
          <Dimmer>
            <OnomyConfirmationModal
              closeModal={() => setApproveModal(false)}
              onConfirm={() => onApprove(swapSellAmount)}
            />
          </Dimmer>
      }
      {
        completedModal && (
          <Dimmer>
            <TransactionCompletedModal
              closeModal={() => setCompletedModal(false)}
              type={completedModal}
              amount={completedAmount}
              result={completedResult}
              previousTx={previousTx}
            />
          </Dimmer>
        )
      }
      {
        failedModal && (<Dimmer>
          <TransactionFailedModal
            closeModal={() => setFailedModal(null)}
            error={failedModal}
          />
        </Dimmer>)
      }
      {
        pendingModal && (
          <Dimmer>
            <PendingModal />
          </Dimmer>
        )
      }
      <ExchangeItem>
        <strong>Buy NOM</strong>
        <Sending>
          <strong>I'm sending</strong>
          <ExchangeInput
            type="text"
            onChange={onBuyNOMTextChange}
            value={swapBuyValue}
          />
          ETH
          <MaxBtn onClick={onEthMax}>Max</MaxBtn>
        </Sending>
        <Receiving>
          <strong>I'm receiving</strong>
          <ReceivingValue>
            {
              BigNumber.isBigNumber(swapBuyResult) ? 
              format18(swapBuyResult).toFixed(4) : 
              ''
            } NOM
          </ReceivingValue>
        </Receiving>
        <div>
          <ExchangeButton onClick={onBuy}>Buy NOM</ExchangeButton>
        </div>
      </ExchangeItem>

      <ExchangeItem>
        <strong>Sell NOM</strong>
        <Sending>
          <strong>I'm sending</strong>
          <ExchangeInput
            type="text"
            onChange={onSellNOMTextChange}
            value={swapSellValue}
          />
          NOM
          <MaxBtn onClick={onNOMMax}>Max</MaxBtn>
        </Sending>
        <Receiving>
          <strong>I'm receiving</strong>
          <ReceivingValue>
            {
              BigNumber.isBigNumber(swapSellResult) ? 
              format18(swapSellResult).toFixed(4) : 
              ''
            } ETH
          </ReceivingValue>
        </Receiving>
        <div>
          {
            NOMallowance > swapSellAmount && NOMbalance > swapSellAmount ? (
              <SellBtn onClick={onSell}>Sell NOM</SellBtn>) : 
                  NOMbalance > swapSellAmount ? (
                    <SellBtn onClick={() => setApproveModal(true)}>Approve</SellBtn>
                  ) : <SellBtn>Not enough NOM</SellBtn>
          }
        </div>
      </ExchangeItem>

      {error && <div>{error}</div>}
    </ExchangeWrapper>
  );
}