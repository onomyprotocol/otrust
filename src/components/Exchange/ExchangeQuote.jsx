import React, { useCallback, useEffect } from 'react';
import { BigNumber } from 'bignumber.js';
import ConfirmTransactionModal from 'components/Modals/components/ConfirmTransactionModal';
import PendingModal from 'components/Modals/components/PendingModal';
import RequestFailedModal from 'components/Modals/components/RequestFailedModal';
import TransactionCompletedModal from 'components/Modals/components/TransactionCompletedModal';
import TransactionFailedModal from 'components/Modals/components/TransactionFailedModal';
import { useWeb3React } from '@web3-react/core';

import { useChain } from 'context/chain/ChainContext';
import { BondingCont } from 'context/chain/contracts';
import { useExchange, useUpdateExchange } from 'context/exchange/ExchangeContext';
import {
  ExchangeItem,
  Sending,
  Receiving,
  ExchangeInput,
  MaxBtn,
  ReceivingValue,
  ExchangeButton,
} from './exchangeStyles';
import { useModal } from 'context/modal/ModalContext';
import NOMButton from 'components/Exchange/NOMButton';
import { format18, parse18 } from 'utils/math';

export default function ExchangeQuote({ strength }) {
  const { strongBalance, weakBalance } = useChain();
  const { handleModal } = useModal();
  const { library } = useWeb3React();

  const bondContract = BondingCont(library);
  // const NOMcontract = NOMCont(library);

  const { askAmount, bidAmount, bidDenom, input, output, strong, weak } = useExchange();

  useEffect(() => {
    console.log('Input: ', input);
    console.log('Output: ', output);
  });

  const { objDispatch, strDispatch } = useUpdateExchange();

  const getAskAmount = useCallback(
    async (askAmountState, bidAmountUpdate, textStrength) => {
      var askAmountUpdate = askAmountState;

      switch (textStrength) {
        case 'strong':
          console.log('Strong: ', bidAmountUpdate.toFixed(0));
          askAmountUpdate = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
          console.log('Pull Strong Ask Amount', askAmountUpdate);
          break;

        case 'weak':
          askAmountUpdate = await bondContract.sellQuoteNOM(bidAmountUpdate.toFixed(0));
          console.log('Pull Weak Ask Amount', askAmountUpdate);
          break;

        default:
          console.error('Denom not set');
      }
      return new BigNumber(askAmountUpdate.toString());
    },
    [bondContract],
  );

  const onApprove = async () => {
    if (bidAmount <= weakBalance) {
      handleModal(<PendingModal type="approving" />);
      // try {
      //   strDispatch({
      //     type: 'status',
      //     value: 'APPROVE',
      //   });

      //   let tx = await NOMcontract.increaseAllowance(bondContract.address, bidAmount.toFixed(0));

      //   tx.wait().then(() => {
      //     handleModal(<TransactionCompletedModal tx={tx} />);
      //   });

      //   strDispatch({
      //     type: 'status',
      //     value: '',
      //   });
      // } catch (e) {
      //   // eslint-disable-next-line no-console
      //   // console.error(e.code, e.message.message);
      //   // alert(e.message)
      //   handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      // }
    } else {
      handleModal(<TransactionFailedModal error={`${weak} Balance too low`} />);
    }
  };

  const submitTrans = useCallback(
    async (slippage, gasPrice) => {
      handleModal(<PendingModal />);
      if (!bidAmount || !askAmount) return;
      try {
        let tx;
        let gasFee;

        switch (bidDenom) {
          case 'strong':
            // Preparing for many tokens / coins
            switch (strong) {
              case 'ETH':
                const gasFeeRaw = await bondContract.estimateGas.buyNOM(askAmount.toFixed(0), slippage.toFixed(0), {
                  value: bidAmount.toFixed(0),
                });

                gasFee = new BigNumber(gasFeeRaw.toString());

                const bidAmountUpdate = bidAmount.minus(gasFee.times(gasPrice));
                const askAmountUpdateRaw = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
                const askAmountUpdate = new BigNumber(askAmountUpdateRaw.toString());

                tx = await bondContract.buyNOM(askAmountUpdate.toFixed(0), slippage.toFixed(0), {
                  value: bidAmountUpdate.toFixed(0),
                  gasPrice: gasPrice.toFixed(0),
                  gasLimit: gasFee.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });

                break;
              default: {
              }
            }
            break;

          case 'weak':
            switch (weak) {
              case 'wNOM':
                tx = await bondContract.sellNOM(bidAmount.toFixed(0), askAmount.toFixed(0), slippage.toFixed(0), {
                  gasPrice: gasPrice.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });
                break;
              default: {
              }
            }
            break;

          default:
            console.log();
        }
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e.code, e.message.message);
        // alert(e.message)
        handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      }
    },
    [askAmount, bidAmount, bidDenom, bondContract, handleModal, strong, weak],
  );

  const onBid = () => {
    switch (true) {
      case bidDenom !== strength:
        handleModal(<RequestFailedModal error="Please enter amount" />);
        break;
      case strength === 'strong' && strongBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      case strength === 'weak' && weakBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      default:
        handleModal(<RequestFailedModal error="Insufficient funds" />);
    }
  };

  const onMax = async () => {
    let strUpdate = new Map();
    strUpdate.set('bidDenom', strength);
    let bidMaxValue = strength === 'strong' ? strongBalance : weakBalance;

    strUpdate.set('input', format18(bidMaxValue).toString());

    let askAmountUpdate;

    try {
      askAmountUpdate = await getAskAmount(askAmount, bidMaxValue, strength);
    } catch (err) {
      if (err) {
        handleModal(<RequestFailedModal error={err.error.message} />);
      }
    }

    let objUpdate = new Map();

    objUpdate = objUpdate.set('askAmount', askAmountUpdate);

    objUpdate = objUpdate.set('bidAmount', bidMaxValue);

    objDispatch({
      type: 'update',
      value: objUpdate,
    });

    strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));

    strDispatch({
      type: 'update',
      value: strUpdate,
    });
  };

  const onTextChange = useCallback(
    async (evt, textStrength) => {
      evt.preventDefault();
      const floatRegExp = new RegExp(/(^(?=.+)(?:[1-9]\d*|0)?(?:\.\d+)?$)|(^\d\.$)/);
      console.log('Component Strength: ', strength);
      console.log('Text Strength: ', textStrength);
      console.log('Bid Denom: ', bidDenom);
      let strUpdate = new Map();
      switch (true) {
        case bidDenom === strength && input === evt.target.value.toString():
          break;
        case evt.target.value === '' || evt.target.value === '.':
          {
            let objUpdate = new Map();

            objUpdate = objUpdate.set('askAmount', new BigNumber(0));

            objUpdate = objUpdate.set('bidAmount', new BigNumber(0));

            objDispatch({
              type: 'update',
              value: objUpdate,
            });
          }

          strUpdate = strUpdate.set('bidDenom', strength);

          strUpdate = strUpdate.set('input', evt.target.value.toString());

          strUpdate = strUpdate.set('output', '');

          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        case floatRegExp.test(evt.target.value.toString()):
          const evttargetvalue = evt.target.value;

          console.log('Input after test', evttargetvalue);

          const bidAmountUpdate = parse18(new BigNumber(parseFloat(evttargetvalue).toString()));

          const inputUpdate = evttargetvalue.toString();

          if (bidDenom !== strength) {
            strUpdate = strUpdate.set('bidDenom', strength);
          }

          var askAmountUpdate;

          try {
            askAmountUpdate = await getAskAmount(askAmount, bidAmountUpdate, textStrength);
          } catch (err) {
            if (err) {
              console.log(err.error.message);
              handleModal(<RequestFailedModal error={err.error.message} />);
            }
            break;
          }

          let objUpdate = new Map();

          objUpdate = objUpdate.set('askAmount', new BigNumber(askAmountUpdate.toString()));

          objUpdate = objUpdate.set('bidAmount', bidAmountUpdate);

          objDispatch({
            type: 'update',
            value: objUpdate,
          });

          strUpdate = strUpdate.set('input', inputUpdate);

          strUpdate = strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));
          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        default:
          handleModal(<RequestFailedModal error="Please enter numbers only. Thank you!" />);
      }
    },
    [askAmount, bidDenom, getAskAmount, handleModal, input, objDispatch, strDispatch, strength],
  );

  return (
    <ExchangeItem>
      <strong>Bid {strength === 'strong' ? strong : weak}</strong>
      <Sending>
        <strong>I'm bidding</strong>
        <ExchangeInput
          type="text"
          onChange={evt => onTextChange(evt, strength)}
          value={bidDenom === strength ? input : ''}
          placeholder="0.00"
        />
        {strength === 'strong' ? strong : weak}
        <MaxBtn onClick={() => onMax()}>Max</MaxBtn>
      </Sending>
      <Receiving>
        <strong>I'm asking</strong>
        <ReceivingValue>
          {strength === bidDenom ? output : ''} {strength === 'strong' ? weak : strong}
        </ReceivingValue>
      </Receiving>
      {strength === 'strong' ? (
        bidDenom === 'weak' ? (
          <ExchangeButton>Input Value</ExchangeButton>
        ) : bidAmount.lte(strongBalance) ? (
          input === '' ? (
            <ExchangeButton>Input Value</ExchangeButton>
          ) : (
            <ExchangeButton onClick={onBid}>Buy {strength === 'strong' ? weak : strong}</ExchangeButton>
          )
        ) : (
          <ExchangeButton>Low {strong} Balance</ExchangeButton>
        )
      ) : (
        <NOMButton onBid={onBid} onApprove={onApprove} />
      )}
    </ExchangeItem>
  );
}
