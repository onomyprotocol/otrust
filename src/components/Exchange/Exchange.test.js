import { cleanup } from '@testing-library/react';

import Exchange from './Exchange';
import { renderWithContext } from 'utils/testing';

describe('Given an Exchange component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', async () => {
      const { asFragment } = renderWithContext(Exchange);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
