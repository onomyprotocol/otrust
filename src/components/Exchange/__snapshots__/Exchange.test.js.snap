// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an Exchange component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM fdCjyu"
  >
    <div
      class="sc-gtsrHT cFutaZ"
    >
      <strong>
        Bid ETH
      </strong>
      <div
        class="sc-dlnjwi ciFpMw"
      >
        <strong>
          I'm bidding
        </strong>
        <input
          class="sc-jSFjdj lkKexS"
          data-testid="exchange-strong-balance-input"
          placeholder="0.00"
          type="text"
          value=""
        />
        ETH
        <button
          class="sc-eCApnc liwdcO"
          data-testid="max-value-button"
        >
          Max
        </button>
      </div>
      <div
        class="sc-hKFxyN jWQojY"
      >
        <strong>
          I'm asking
        </strong>
        <span
          class="sc-gKAaRy dHKXDL"
          data-testid="exchange-weak-balance"
        >
           NOM
        </span>
      </div>
      <button
        class="sc-iCoGMd SknTJ"
      >
        Low ETH Balance
      </button>
    </div>
    <div
      class="sc-gtsrHT cFutaZ"
    >
      <strong>
        Bid NOM
      </strong>
      <div
        class="sc-dlnjwi ciFpMw"
      >
        <strong>
          I'm bidding
        </strong>
        <input
          class="sc-jSFjdj lkKexS"
          data-testid="exchange-strong-balance-input"
          placeholder="0.00"
          type="text"
          value=""
        />
        NOM
        <button
          class="sc-eCApnc liwdcO"
          data-testid="max-value-button"
        >
          Max
        </button>
      </div>
      <div
        class="sc-hKFxyN jWQojY"
      >
        <strong>
          I'm asking
        </strong>
        <span
          class="sc-gKAaRy dHKXDL"
          data-testid="exchange-weak-balance"
        >
           ETH
        </span>
      </div>
      <button
        class="sc-iCoGMd sc-fujyAs SknTJ hBWDAL"
      >
         Input Value 
      </button>
    </div>
  </div>
</DocumentFragment>
`;
